js/actions/tests/loadSourceText.js

```diff
@@ -23,20 +29,111 @@ const mockThreadClient = {
+    return this.request;
+  }
+};
+
 describe("loadSourceText", () => {
-  beforeEach(function() {
-    this.store = createStore(mockThreadClient);
+  describe("loading one source text", function() {
+    beforeEach(function(done) {
+      this.store = createStore(simpleMockThreadClient);
+      this.store.dispatch(actions.loadSourceText({ actor: "foo1" }))
+        .then(() => done());
+    });
+
+    it("Store has the source text", function() {
+      const fooSourceText = queries.getSourceText(
+                              this.store.getState(), "foo1");
+      expect(fooSourceText.get("text")).to.equal(sourceText.foo1.source);
+    });
   });
 
-  describe("loading one source text", function() {
+  describe("loading two different sources", function() {
     beforeEach(function(done) {
+      this.store = createStore(simpleMockThreadClient);
       this.store.dispatch(actions.loadSourceText({ actor: "foo1" }))
+        .then(() => {
+          return this.store.dispatch(actions.loadSourceText({ actor: "foo2" }));
+        })
         .then(() => done());
     });
 
+    it("Store has the first source text", function() {
+      const fooSourceText = queries.getSourceText(
+                              this.store.getState(), "foo1");
+      expect(fooSourceText.get("text")).to.equal(sourceText.foo1.source);
+    });
+
+    it("Store has the second source text", function() {
+      const fooSourceText = queries.getSourceText(
+                              this.store.getState(), "foo2");
+      expect(fooSourceText.get("text")).to.equal(sourceText.foo2.source);
+    });
+  });
+
+  describe("loading a source twice", function() {
+    beforeEach(function(done) {
+      const store = createStore(simpleMockThreadClient);
+      this.store = store;
+
+      Task.spawn(function* () {
+        yield store.dispatch(actions.loadSourceText({ actor: "foo1" }));
+        yield store.dispatch(actions.loadSourceText({ actor: "foo1" }));
+        done();
+      });
+    });
+
     it("Store has the source text", function() {
-      const fooSourceText = queries.getSourceText(this.store.getState(), "foo1");
+      const fooSourceText = queries.getSourceText(
+                              this.store.getState(), "foo1");
       expect(fooSourceText.get("text")).to.equal(sourceText.foo1.source);
     });
   });
+
+  describe("source is loading", function() {
+    beforeEach(function(done) {
+      this.store = createStore(deferredMockThreadClient);
+      this.store.dispatch(actions.loadSourceText({ actor: "foo1" }));
+      // We're intentionally leaving the source promise pending
+      done();
+    });
+
+    it("Store has a loading source text", function() {
+      const fooSourceText = queries.getSourceText(
+                              this.store.getState(), "foo1");
+      expect(fooSourceText.get("loading")).to.equal(true);
+    });
+  });
+
+  describe("source failed to load", function() {
+    beforeEach(function(done) {
```

Comment:
Why use `beforeEach` when you only have one test?

It seems like there should be one top-level describe and then several "its". The describe should be just take the string "sources" and then the `it` blocks are something like `it("fails to load", ...)`. At least I thought that's how mocha worked (`describe` specifies the thing you are testing, and `it` is the action).